hikari/impl/event_manager_base.py

```diff
@@ -223,9 +223,13 @@ def dispatch(self, event: event_manager.EventT_inv) -> asyncio.Future[typing.Any
                 for callback in self._listeners[cls]:
                     tasks.append(self._invoke_callback(callback, event))
 
-            if cls in self._waiters:
-                waiter_set = self._waiters[cls]
-                for predicate, future in tuple(waiter_set):
+            waiter_set = self._waiters.pop(cls, None)
+            if not waiter_set:
+                continue
+
+            for waiter in waiter_set:
+                predicate, future = waiter
+                try:
```

Comment:
Always popping the waiters will not have the effect we want. The way this currently works is that the waiters wont be removed if the result of the predicate is not trully.

This code will remove the waiter always, regardless if the future gets completed or if the event is the one that we want.

Furthermore, the bug that you are trying to fix lies somewhere else as, if we cleanup everything correctly, there should not be an invalid state

Reply:
ah ok